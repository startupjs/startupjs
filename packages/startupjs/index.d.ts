export * from '@startupjs/react-sharedb'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline
export function css (): void
export function styl (): void
