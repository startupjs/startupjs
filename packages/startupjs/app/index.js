module.exports = require('@startupjs/app')
