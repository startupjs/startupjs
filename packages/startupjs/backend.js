export { default } from '@startupjs/backend'
export * from '@startupjs/backend'
