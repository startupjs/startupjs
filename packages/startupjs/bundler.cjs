throw new Error(`
[startupjs/bundler] ERROR! StartupJS bundler api has changed.

Now instead of using general \`require('startupjs/bundler.cjs')\` you have to require
a specific file you need.

Here is what you need to change:

1. \`require('startupjs/bundler.cjs').webpackWebConfig\`
    v v v
   \`require('startupjs/bundler/webpack.web.config.cjs')\`

2. \`require('startupjs/bundler.cjs').webpackServerConfig\`
    v v v
   \`require('startupjs/bundler/webpack.server.config.cjs')\`

3. \`require('startupjs/bundler.cjs').babelConfig\`
    v v v
   \`require('startupjs/bundler/babel.config.cjs')\`

4. \`require('startupjs/bundler.cjs').metroConfig\`
    v v v
   \`require('startupjs/bundler/metro.config.cjs')\`

5. \`require('startupjs/bundler.cjs').rnConfig\`
    v v v
   \`require('startupjs/bundler/react-native.config.cjs')\`

6. \`require('startupjs/bundler.cjs').viteConfig\`
    v v v
   \`require('startupjs/bundler/vite.config.cjs')\`

Additionally, if you did override .js/.jsx compilation in your webpack.web.config.cjs, you
might have used \`getJsxRule\` function there. In which case you should now get it from helpers:

7. \`const { getJsxRule } = require('startupjs/bundler.cjs')\`
    v v v
   \`const { getJsxRule } = require('startupjs/bundler/helpers.cjs')\`
`)
