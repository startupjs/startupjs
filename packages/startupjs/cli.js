#!/usr/bin/env node
'use strict'

import cli from '@startupjs/cli'
import { fileURLToPath } from 'url'
import path, { dirname } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const templatesPath = path.join(__dirname, 'templates')

cli.run({ templatesPath })
