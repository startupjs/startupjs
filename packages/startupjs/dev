#!/bin/sh
# shellcheck disable=SC2059

# Exit on first failed command
set -e

# Run startupjs project using only Docker
#
# Version: 0.1.0
#
# Suports Mac, Linux, Windows WSL2
#
# Usage:
#   ./dev - universal command to run the main terminal or spawn additional terminals.
#           It tries to do the following:
#           1. Run the dev docker image of the same version of startupjs as in package.json.
#           2. If dev image is already running -- exec into it.

# Use current folder name as the name for container and the image by default
# ref: https://stackoverflow.com/a/43919044
a="/$0"; a=${a%/*}; a=${a#/}; a=${a:-.}; SCRIPT_DIR=$(cd "$a"; pwd)
FOLDER_NAME=$( basename "$SCRIPT_DIR" )

IMAGE_NAME="startupjs/dev"
CONTAINER_NAME="$FOLDER_NAME"

# Get startupjs version from package.json
STARTUPJS_VERSION=$( [ -f ./package.json ] && (grep -m 1 '"startupjs"' < ./package.json | sed -n 's/^[^"]*"[^"]*"[^"]*"//p' | sed -n 's/"[^"]*$//p') || echo '' )
STARTUPJS_VERSION=${STARTUPJS_VERSION:-latest}
# Remove leading ^ or ~ and remove the last patch number.
# This way ^0.40.2 will be converted to 0.40
# TODO: if yarn.lock exists, get the exact patch version from there
STARTUPJS_VERSION=$( echo "$STARTUPJS_VERSION" | sed -n 's/^[\^\~]*//p' | sed -n 's/\.*[0-9]*$//p' )

# Helper constants
ROOT="${PWD}"
# folder to store the MongoDB data
DB_FOLDER="${ROOT}/.mongo"
OS_NAME=$( uname -s | tr A-Z a-z )
# on Mac OS we use a faster driver for volume mounts
VOLUME_MODE=$( [ "$OS_NAME" = "darwin" ] && echo ":delegated" || echo "" )

# Colors
COLOR_NONE='\033[0m'
COLOR_LIGHT_GREEN='\033[1;32m'
COLOR_LIGHT_RED='\033[1;31m'

main () {
  if docker ps | grep "\s${CONTAINER_NAME}$"; then
    printf "${COLOR_LIGHT_GREEN}EXEC into the running container '${CONTAINER_NAME}'...${COLOR_NONE}\n"
    _exec
  else
    printf "${COLOR_LIGHT_GREEN}RUN a new dev container '${CONTAINER_NAME}'...${COLOR_NONE}\n"
    _cleanup
    _run
  fi
}

_exec () {
  printf "${COLOR_LIGHT_GREEN}You are now inside the dev environment. Run your commands here:${COLOR_NONE}\n"
  docker exec -it \
    -e "CONTAINER_NAME=${CONTAINER_NAME}" \
    "${CONTAINER_NAME}" \
    bash
}

_run () {
  PORT="${PORT:-$( _find_open_PORT )}"
  DEV_PORT="${DEV_PORT:-$( _find_open_DEV_PORT )}"
  if [ "$PORT" != "3000" ]; then
    printf "${COLOR_LIGHT_RED}IMPORTANT: Default port 3000 was occupied. Your app got assigned the port ${PORT}${COLOR_NONE}!\n"
  fi
  printf "${COLOR_LIGHT_GREEN}You are now inside the dev environment. Run your commands here:${COLOR_NONE}\n"
  echo "CONTAINER_NAME: $CONTAINER_NAME"
  echo "PORT: $PORT"
  echo "DEV_PORT: $DEV_PORT"
  echo "ROOT: $ROOT"
  echo "VOLUME_MODE: $VOLUME_MODE"
  echo "HOME: $HOME"
  echo "DB_FOLDER: $DB_FOLDER"
  echo "IMAGE_NAME: $IMAGE_NAME"
  echo "STARTUPJS_VERSION: $STARTUPJS_VERSION"
  docker run --rm -it \
    --name "${CONTAINER_NAME}" \
    --pull=always \
    -e "CONTAINER_NAME=${CONTAINER_NAME}" \
    -e "DOCKER_HOST_PORT=${PORT}" \
    --ulimit nofile=65535:65535 \
    -p "${PORT}:3000" \
    -p "${DEV_PORT}:3010" \
    -v "${ROOT}:/ws${VOLUME_MODE}" \
    -v "${HOME}/.ssh:/root/.ssh" \
    -v "${HOME}/.gitconfig:/root/.gitconfig" \
    -v "${DB_FOLDER}:/data/db${VOLUME_MODE}" \
    "${IMAGE_NAME}:${STARTUPJS_VERSION}"
}

# Keep only 5 latest startupjs images you pulled. Remove older ones
# ref: https://stackoverflow.com/a/40892314
_cleanup () {
  # shellcheck disable=SC2046
  docker rmi $(docker images -q ${IMAGE_NAME} | tail -n +6) || true
}

# a bit hardcody way to get available port for the app. Start with 3000 and go up to 3009
_find_open_PORT () {
  for iterator in 0 1 2 3 4 5 6 7 8 9; do
    a_port="300${iterator}"
    if ! lsof -n -i tcp:$a_port; then
      echo $a_port
      break
    fi
  done
}

# the same hardcody way as above to get available port for webpack devserver. Start with 3010 and go up to 3019
_find_open_DEV_PORT () {
  for iterator in 0 1 2 3 4 5 6 7 8 9; do
    a_port="301${iterator}"
    if ! lsof -n -i tcp:$a_port; then
      echo $a_port
      break
    fi
  done
}

main "$@"; exit
