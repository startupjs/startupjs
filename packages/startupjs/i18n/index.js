export * from '@startupjs/i18n'
