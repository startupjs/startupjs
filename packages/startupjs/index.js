export * from '@startupjs/react-sharedb'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline
export function css () {}
export function styl () {}
export { default as t } from '@startupjs/i18n/client/t.js'
