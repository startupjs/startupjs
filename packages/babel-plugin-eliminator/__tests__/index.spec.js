const pluginTester = require('babel-plugin-tester').default
const plugin = require('../index')
const { name: pluginName } = require('../package.json')

const LONGER_EXAMPLE = /* js */`
  import usedByFoo from 'used-by-foo'
  import usedByDefault from 'used-by-default'
  import usedByBar from 'used-by-bar'

  const varInFoo = 'var-in-foo'
  const varInDefault = 'var-in-default'
  const varInBar = 'var-in-bar'

  export const foo = () => {
    return usedByFoo(varInFoo)
  }

  export default () => {
    return usedByDefault(varInDefault)
  }

  export function bar () {
    return usedByBar(varInBar)
  }
`

pluginTester({
  plugin,
  pluginName,
  snapshot: true,
  pluginOptions: {
    removeExports: ['preload', 'staticPreload']
  },
  babelOptions: {
    filename: '/ws/dummy/index.js'
  },
  tests: {
    'removes export function declaration. 1': /* js */`
      export function preload() {
        return {}
      }
    `,
    'removes export function declaration. 2': /* js */`
      export function staticPreload() {
        return {}
      }
    `,
    'removes references that are only used in server exports': /* js */`
      var readFile = require('fs').readFile
      import foo from 'foo'
      // This is kept since it's not used by ssr exports
      import { getStaticProps } from './'
      const a = 1
      function bar() {

      }
      var shouldHeep = 2
      var alsoShouldKeep =3
      console.log(alsoShouldKepp)
      export function preload() {
        return {a, bar, foo, readFile: readFile.toString()}
      }
    `,
    'removes export variable declaration': /* js */`
      export var preload= function() {
        return {}
      }
    `,
    'removes re-exports': /* js */`
      export {preload} from './'
    `,
    'keeps other exports': /* js */`
      export const a = 1
      export {preload} from './'
    `,
    'removes destructuring assignment (array)': /* js */`
      const [a,b,c] = d
      const e = d[0]
      export const preload = () => {
        console.log(a,b,c,e)
      }
    `,
    'longer example. Remove `foo`': {
      pluginOptions: {
        removeExports: ['foo']
      },
      code: LONGER_EXAMPLE
    },
    'longer example. Remove `foo`, `default`': {
      pluginOptions: {
        removeExports: ['foo', 'default']
      },
      code: LONGER_EXAMPLE
    },
    'throws error if both `removeExports` and `keepExports` are specified': {
      pluginOptions: {
        removeExports: ['foo'],
        keepExports: ['bar']
      },
      error: true,
      code: /* js */`
        export const foo = () => {}
        export function bar () {}
      `
    },
    'longer example. Keep `foo`': {
      pluginOptions: {
        keepExports: ['foo']
      },
      code: LONGER_EXAMPLE
    },
    'longer example. Keep `foo`, `default`': {
      pluginOptions: {
        keepExports: ['foo', 'default']
      },
      code: LONGER_EXAMPLE
    },
    'object in magic function. Only parse magic function': {
      pluginOptions: {
        trimObjects: [{
          functionName: 'createProject',
          magicImports: ['startupjs/registry', '@startupjs/registry'],
          targetObjectJsonPath: '$.plugins.*',
          ensureOnlyKeys: ['client', 'isomorphic', 'server', 'build'],
          keepKeys: ['client']
        }]
      },
      code: /* js */`
        import { createProject2 } from 'startupjs/registry'
        export default createProject2({
          plugins: {
            'serve-static-promo': {
              client: {
                redirectUrl: '/promo',
              },
              server: {
                testServer: 'hello server',
              }
            }
          }
        })
      `
    },
    'object in magic function. Throw error if there are keys other than `ensureOnlyKeys`': {
      pluginOptions: {
        trimObjects: [{
          functionName: 'createProject',
          magicImports: ['startupjs/registry', '@startupjs/registry'],
          targetObjectJsonPath: '$.plugins.*',
          ensureOnlyKeys: ['client', 'isomorphic', 'server', 'build'],
          keepKeys: ['client']
        }]
      },
      error: true,
      code: /* js */`
        import { createProject } from 'startupjs/registry'
        export default createProject({
          plugins: {
            'serve-static-promo': {
              client: {
                redirectUrl: '/promo'
              }
            },
            permissions: {
              client: {
                roles: ['admin', 'user']
              },
              magic: {
                testMagic: 'hello magic'
              }
            }
          }
        })
      `
    },
    // Removing keys from object within magic function call
    'object in magic function. Keep keys `client` and `isomorphic`': {
      pluginOptions: {
        trimObjects: [{
          functionName: 'createProject',
          magicImports: ['startupjs/registry', '@startupjs/registry'],
          targetObjectJsonPath: '$.plugins.*',
          ensureOnlyKeys: ['client', 'isomorphic', 'server', 'build'],
          keepKeys: ['client', 'isomorphic']
        }]
      },
      code: /* js */`
        import { createProject } from 'startupjs/registry'
        import { clientLib, clientLib2 } from 'client-lib'
        import { serverLib, serverLib2 } from 'server-lib'
        import buildLib from 'build-lib'
        import isomorphicLib from 'isomorphic-lib'

        const clientVar = clientLib()
        const clientVar2 = clientLib2()
        const serverVar = serverLib()
        const serverVar2 = serverLib2()
        const buildVar = buildLib()
        const isomorphicVar = isomorphicLib()

        export default createProject({
          plugins: {
            'serve-static-promo': {
              client: {
                redirectUrl: '/promo',
                testClient: 'hello client',
                clientVar
              },
              server: {
                testServer: 'hello server',
                serverVar
              },
              build: {
                testBuild: 'hello build',
                buildVar
              },
              isomorphic: {
                testIsomorphic: 'hello isomorphic',
                isomorphicVar
              }
            },
            permissions: {
              client: {
                testClient: 'permissions client',
                clientVar2
              },
              server: {
                testServer: 'permissions server',
                roles: ['admin', 'user'],
                serverVar2
              },
              build: {
                testBuild: 'permissions build',
                buildVar
              },
              isomorphic: {
                testIsomorphic: 'permissions isomorphic',
                isomorphicVar
              }
            }
          }
        })
      `
    },
    // Removing keys from object within default export of the magic filename
    'object in default export of magic filename. Keep keys `client` and `isomorphic`': {
      pluginOptions: {
        trimObjects: [{
          magicFilenameRegex: /startupjs\.config\.js/,
          magicExport: 'default',
          targetObjectJsonPath: '$.modules.*',
          ensureOnlyKeys: ['client', 'isomorphic', 'server', 'build'],
          keepKeys: ['client', 'isomorphic']
        }, {
          magicFilenameRegex: /startupjs\.config\.js/,
          magicExport: 'default',
          targetObjectJsonPath: '$.plugins.*',
          ensureOnlyKeys: ['client', 'isomorphic', 'server', 'build'],
          keepKeys: ['client', 'isomorphic']
        }, {
          magicFilenameRegex: /startupjs\.config\.js/,
          magicExport: 'default',
          targetObjectJsonPath: '$',
          ensureOnlyKeys: ['plugins', 'modules', 'client', 'isomorphic', 'server', 'build'],
          keepKeys: ['plugins', 'modules', 'client', 'isomorphic']
        }]
      },
      babelOptions: {
        filename: '/ws/dummy/startupjs.config.js'
      },
      code: /* js */`
        import { clientLib, clientLib2 } from 'client-lib'
        import { serverLib, serverLib2 } from 'server-lib'
        import buildLib from 'build-lib'
        import isomorphicLib from 'isomorphic-lib'

        const clientVar = clientLib()
        const clientVar2 = clientLib2()
        const serverVar = serverLib()
        const serverVar2 = serverLib2()
        const buildVar = buildLib()
        const isomorphicVar = isomorphicLib()

        export default {
          isomorphic: {
            allowUnusedPlugins: true,
            server: true
          },
          server: {
            init: options => ({
              api: expressApp => {
                expressApp.get('/hello', (req, res) => {
                  res.send('Hello from server')
                })
              }
            })
          },
          plugins: {
            'serve-static-promo': {
              client: {
                redirectUrl: '/promo',
                testClient: 'hello client',
                clientVar
              },
              server: {
                testServer: 'hello server',
                serverVar
              },
              build: {
                testBuild: 'hello build',
                buildVar
              },
              isomorphic: {
                testIsomorphic: 'hello isomorphic',
                isomorphicVar
              }
            },
            permissions: {
              client: {
                testClient: 'permissions client',
                clientVar2
              },
              server: {
                testServer: 'permissions server',
                roles: ['admin', 'user'],
                serverVar2
              },
              build: {
                testBuild: 'permissions build',
                buildVar
              },
              isomorphic: {
                testIsomorphic: 'permissions isomorphic',
                isomorphicVar
              }
            }
          }
        }
      `
    },
    // Removing keys from object within default export of the magic filename
    'transform function calls': {
      pluginOptions: {
        transformFunctionCalls: [{
          // direct named exports of aggregation() within model/*.js files
          // are replaced with aggregationHeader() calls.
          // 'collection' is the filename without extension
          // 'name' is the direct named export const name
          //
          // Example:
          //
          //   // in model/games.js
          //   export const $$byGameId = aggregation(({ gameId }) => ({ gameId }))
          //
          // will be replaced with:
          //
          //   __aggregationHeader({ collection: 'games', name: '$$byGameId' })
          //
          functionName: 'aggregation',
          magicImports: ['@startupjs/orm', 'startupjs/orm'],
          requirements: {
            directNamedExportedAsConst: true
          },
          throwIfRequirementsNotMet: true, // TODO: remove this when the next transformation is implemented
          replaceWith: {
            newFunctionNameFromSameImport: '__aggregationHeader',
            newCallArgumentsTemplate: `[
              {
                collection: %%filenameWithoutExtension%%,
                name: %%directNamedExportConstName%%
              }
            ]`
          }
        }, {
          // remove accessControl() calls (replace with undefined)
          functionName: 'accessControl',
          magicImports: ['@startupjs/orm', 'startupjs/orm'],
          replaceWith: {
            remove: true // replace the whole function call with undefined
          }
        }]
      },
      babelOptions: {
        filename: '/ws/dummy/model/games.js'
      },
      code: /* js */`
        import { aggregation, BaseModel, accessControl } from 'startupjs/orm'
        import { getAppName, getRoleId, DEFAULT_USER_ID } from 'server-lib'

        const appName = getAppName()
        const adminRoleId = getRoleId('admin')

        export const access = accessControl({
          create: (doc, { session }) => session.roleId === adminRoleId,
          read: () => true,
          update: (doc, { session }) => session.roleId === adminRoleId || !doc.readonly,
          delete: (doc, { session }) => session.roleId === adminRoleId
        })

        export const $$createdByUser = aggregation(({ userId = DEFAULT_USER_ID }) => ({
          userId,
          $sort: { createdAt: -1 },
          appName
        }))

        export default class GamesModel extends BaseModel {
          async addNew () {
            await this.add({ name: 'New Game' })
          }
        }
      `
    }
  }
})
