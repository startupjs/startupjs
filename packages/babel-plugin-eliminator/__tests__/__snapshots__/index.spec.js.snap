// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@startupjs/babel-plugin-eliminator keeps other exports: keeps other exports 1`] = `

export const a = 1
export {preload} from './'

      â†“ â†“ â†“ â†“ â†“ â†“

export const a = 1;
export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator longer example. Keep \`foo\`, \`default\`: longer example. Keep \`foo\`, \`default\` 1`] = `

import usedByFoo from 'used-by-foo'
import usedByDefault from 'used-by-default'
import usedByBar from 'used-by-bar'

const varInFoo = 'var-in-foo'
const varInDefault = 'var-in-default'
const varInBar = 'var-in-bar'

export const foo = () => {
  return usedByFoo(varInFoo)
}

export default () => {
  return usedByDefault(varInDefault)
}

export function bar () {
  return usedByBar(varInBar)
}

      â†“ â†“ â†“ â†“ â†“ â†“

import usedByFoo from "used-by-foo";
import usedByDefault from "used-by-default";
const varInFoo = "var-in-foo";
const varInDefault = "var-in-default";
export const foo = () => {
  return usedByFoo(varInFoo);
};
export default () => {
  return usedByDefault(varInDefault);
};
export var bar = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator longer example. Keep \`foo\`: longer example. Keep \`foo\` 1`] = `

import usedByFoo from 'used-by-foo'
import usedByDefault from 'used-by-default'
import usedByBar from 'used-by-bar'

const varInFoo = 'var-in-foo'
const varInDefault = 'var-in-default'
const varInBar = 'var-in-bar'

export const foo = () => {
  return usedByFoo(varInFoo)
}

export default () => {
  return usedByDefault(varInDefault)
}

export function bar () {
  return usedByBar(varInBar)
}

      â†“ â†“ â†“ â†“ â†“ â†“

import usedByFoo from "used-by-foo";
const varInFoo = "var-in-foo";
export const foo = () => {
  return usedByFoo(varInFoo);
};
export default undefined;
export var bar = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator longer example. Remove \`foo\`, \`default\`: longer example. Remove \`foo\`, \`default\` 1`] = `

import usedByFoo from 'used-by-foo'
import usedByDefault from 'used-by-default'
import usedByBar from 'used-by-bar'

const varInFoo = 'var-in-foo'
const varInDefault = 'var-in-default'
const varInBar = 'var-in-bar'

export const foo = () => {
  return usedByFoo(varInFoo)
}

export default () => {
  return usedByDefault(varInDefault)
}

export function bar () {
  return usedByBar(varInBar)
}

      â†“ â†“ â†“ â†“ â†“ â†“

import usedByBar from "used-by-bar";
const varInBar = "var-in-bar";
export var foo = undefined;
export default undefined;
export function bar() {
  return usedByBar(varInBar);
}


`;

exports[`@startupjs/babel-plugin-eliminator longer example. Remove \`foo\`: longer example. Remove \`foo\` 1`] = `

import usedByFoo from 'used-by-foo'
import usedByDefault from 'used-by-default'
import usedByBar from 'used-by-bar'

const varInFoo = 'var-in-foo'
const varInDefault = 'var-in-default'
const varInBar = 'var-in-bar'

export const foo = () => {
  return usedByFoo(varInFoo)
}

export default () => {
  return usedByDefault(varInDefault)
}

export function bar () {
  return usedByBar(varInBar)
}

      â†“ â†“ â†“ â†“ â†“ â†“

import usedByDefault from "used-by-default";
import usedByBar from "used-by-bar";
const varInDefault = "var-in-default";
const varInBar = "var-in-bar";
export var foo = undefined;
export default () => {
  return usedByDefault(varInDefault);
};
export function bar() {
  return usedByBar(varInBar);
}


`;

exports[`@startupjs/babel-plugin-eliminator object in default export of magic filename. Keep keys \`client\` and \`isomorphic\`: object in default export of magic filename. Keep keys \`client\` and \`isomorphic\` 1`] = `

import { clientLib, clientLib2 } from 'client-lib'
import { serverLib, serverLib2 } from 'server-lib'
import buildLib from 'build-lib'
import isomorphicLib from 'isomorphic-lib'

const clientVar = clientLib()
const clientVar2 = clientLib2()
const serverVar = serverLib()
const serverVar2 = serverLib2()
const buildVar = buildLib()
const isomorphicVar = isomorphicLib()

export default {
  isomorphic: {
    allowUnusedPlugins: true,
    server: true
  },
  server: {
    init: options => ({
      api: expressApp => {
        expressApp.get('/hello', (req, res) => {
          res.send('Hello from server')
        })
      }
    })
  },
  plugins: {
    'serve-static-promo': {
      client: {
        redirectUrl: '/promo',
        testClient: 'hello client',
        clientVar
      },
      server: {
        testServer: 'hello server',
        serverVar
      },
      build: {
        testBuild: 'hello build',
        buildVar
      },
      isomorphic: {
        testIsomorphic: 'hello isomorphic',
        isomorphicVar
      }
    },
    permissions: {
      client: {
        testClient: 'permissions client',
        clientVar2
      },
      server: {
        testServer: 'permissions server',
        roles: ['admin', 'user'],
        serverVar2
      },
      build: {
        testBuild: 'permissions build',
        buildVar
      },
      isomorphic: {
        testIsomorphic: 'permissions isomorphic',
        isomorphicVar
      }
    }
  }
}

      â†“ â†“ â†“ â†“ â†“ â†“

import { clientLib, clientLib2 } from "client-lib";
import isomorphicLib from "isomorphic-lib";
const clientVar = clientLib();
const clientVar2 = clientLib2();
const isomorphicVar = isomorphicLib();
export default {
  isomorphic: {
    allowUnusedPlugins: true,
    server: true,
  },
  plugins: {
    "serve-static-promo": {
      client: {
        redirectUrl: "/promo",
        testClient: "hello client",
        clientVar,
      },
      isomorphic: {
        testIsomorphic: "hello isomorphic",
        isomorphicVar,
      },
    },
    permissions: {
      client: {
        testClient: "permissions client",
        clientVar2,
      },
      isomorphic: {
        testIsomorphic: "permissions isomorphic",
        isomorphicVar,
      },
    },
  },
};


`;

exports[`@startupjs/babel-plugin-eliminator object in magic function. Keep keys \`client\` and \`isomorphic\`: object in magic function. Keep keys \`client\` and \`isomorphic\` 1`] = `

import { createProject } from 'startupjs/registry'
import { clientLib, clientLib2 } from 'client-lib'
import { serverLib, serverLib2 } from 'server-lib'
import buildLib from 'build-lib'
import isomorphicLib from 'isomorphic-lib'

const clientVar = clientLib()
const clientVar2 = clientLib2()
const serverVar = serverLib()
const serverVar2 = serverLib2()
const buildVar = buildLib()
const isomorphicVar = isomorphicLib()

export default createProject({
  plugins: {
    'serve-static-promo': {
      client: {
        redirectUrl: '/promo',
        testClient: 'hello client',
        clientVar
      },
      server: {
        testServer: 'hello server',
        serverVar
      },
      build: {
        testBuild: 'hello build',
        buildVar
      },
      isomorphic: {
        testIsomorphic: 'hello isomorphic',
        isomorphicVar
      }
    },
    permissions: {
      client: {
        testClient: 'permissions client',
        clientVar2
      },
      server: {
        testServer: 'permissions server',
        roles: ['admin', 'user'],
        serverVar2
      },
      build: {
        testBuild: 'permissions build',
        buildVar
      },
      isomorphic: {
        testIsomorphic: 'permissions isomorphic',
        isomorphicVar
      }
    }
  }
})

      â†“ â†“ â†“ â†“ â†“ â†“

import { createProject } from "startupjs/registry";
import { clientLib, clientLib2 } from "client-lib";
import isomorphicLib from "isomorphic-lib";
const clientVar = clientLib();
const clientVar2 = clientLib2();
const isomorphicVar = isomorphicLib();
export default createProject({
  plugins: {
    "serve-static-promo": {
      client: {
        redirectUrl: "/promo",
        testClient: "hello client",
        clientVar,
      },
      isomorphic: {
        testIsomorphic: "hello isomorphic",
        isomorphicVar,
      },
    },
    permissions: {
      client: {
        testClient: "permissions client",
        clientVar2,
      },
      isomorphic: {
        testIsomorphic: "permissions isomorphic",
        isomorphicVar,
      },
    },
  },
});


`;

exports[`@startupjs/babel-plugin-eliminator object in magic function. Only parse magic function: object in magic function. Only parse magic function 1`] = `

import { createProject2 } from 'startupjs/registry'
export default createProject2({
  plugins: {
    'serve-static-promo': {
      client: {
        redirectUrl: '/promo',
      },
      server: {
        testServer: 'hello server',
      }
    }
  }
})

      â†“ â†“ â†“ â†“ â†“ â†“

import { createProject2 } from "startupjs/registry";
export default createProject2({
  plugins: {
    "serve-static-promo": {
      client: {
        redirectUrl: "/promo",
      },
      server: {
        testServer: "hello server",
      },
    },
  },
});


`;

exports[`@startupjs/babel-plugin-eliminator object in magic function. Throw error if there are keys other than \`ensureOnlyKeys\`: object in magic function. Throw error if there are keys other than \`ensureOnlyKeys\` 1`] = `

import { createProject } from 'startupjs/registry'
export default createProject({
  plugins: {
    'serve-static-promo': {
      client: {
        redirectUrl: '/promo'
      }
    },
    permissions: {
      client: {
        roles: ['admin', 'user']
      },
      magic: {
        testMagic: 'hello magic'
      }
    }
  }
})

      â†“ â†“ â†“ â†“ â†“ â†“

SyntaxError: /ws/dummy/index.js: 
createProject(): key is not listed in ensureOnlyKeys: 'magic'.
You can only use keys listed in \`ensureOnlyKeys\`:
  - client
  - isomorphic
  - server
  - build
[0m [90m 11 |[39m         roles[33m:[39m [[32m'admin'[39m[33m,[39m [32m'user'[39m][0m
[0m [90m 12 |[39m       }[33m,[39m[0m
[0m[31m[1m>[22m[39m[90m 13 |[39m       magic[33m:[39m {[0m
[0m [90m    |[39m       [31m[1m^[22m[39m[0m
[0m [90m 14 |[39m         testMagic[33m:[39m [32m'hello magic'[39m[0m
[0m [90m 15 |[39m       }[0m
[0m [90m 16 |[39m     }[0m

`;

exports[`@startupjs/babel-plugin-eliminator removes destructuring assignment (array): removes destructuring assignment (array) 1`] = `

const [a,b,c] = d
const e = d[0]
export const preload = () => {
  console.log(a,b,c,e)
}

      â†“ â†“ â†“ â†“ â†“ â†“

export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator removes export function declaration. 1: removes export function declaration. 1 1`] = `

export function preload() {
  return {}
}

      â†“ â†“ â†“ â†“ â†“ â†“

export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator removes export function declaration. 2: removes export function declaration. 2 1`] = `

export function staticPreload() {
  return {}
}

      â†“ â†“ â†“ â†“ â†“ â†“

export var staticPreload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator removes export variable declaration: removes export variable declaration 1`] = `

export var preload= function() {
  return {}
}

      â†“ â†“ â†“ â†“ â†“ â†“

export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator removes re-exports: removes re-exports 1`] = `

export {preload} from './'

      â†“ â†“ â†“ â†“ â†“ â†“

export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator removes references that are only used in server exports: removes references that are only used in server exports 1`] = `

var readFile = require('fs').readFile
import foo from 'foo'
// This is kept since it's not used by ssr exports
import { getStaticProps } from './'
const a = 1
function bar() {

}
var shouldHeep = 2
var alsoShouldKeep =3
console.log(alsoShouldKepp)
export function preload() {
  return {a, bar, foo, readFile: readFile.toString()}
}

      â†“ â†“ â†“ â†“ â†“ â†“

// This is kept since it's not used by ssr exports
import { getStaticProps } from "./";
var shouldHeep = 2;
var alsoShouldKeep = 3;
console.log(alsoShouldKepp);
export var preload = undefined;


`;

exports[`@startupjs/babel-plugin-eliminator throws error if both \`removeExports\` and \`keepExports\` are specified: throws error if both \`removeExports\` and \`keepExports\` are specified 1`] = `

export const foo = () => {}
export function bar () {}

      â†“ â†“ â†“ â†“ â†“ â†“

Error: /ws/dummy/index.js: You cannot specify both \`removeExports\` and \`keepExports\`

`;

exports[`@startupjs/babel-plugin-eliminator transform function calls: transform function calls 1`] = `

import { aggregation, BaseModel, accessControl } from 'startupjs/orm'
import { getAppName, getRoleId, DEFAULT_USER_ID } from 'server-lib'

const appName = getAppName()
const adminRoleId = getRoleId('admin')

export const access = accessControl({
  create: (doc, { session }) => session.roleId === adminRoleId,
  read: () => true,
  update: (doc, { session }) => session.roleId === adminRoleId || !doc.readonly,
  delete: (doc, { session }) => session.roleId === adminRoleId
})

export const $$createdByUser = aggregation(({ userId = DEFAULT_USER_ID }) => ({
  userId,
  $sort: { createdAt: -1 },
  appName
}))

export default class GamesModel extends BaseModel {
  async addNew () {
    await this.add({ name: 'New Game' })
  }
}

      â†“ â†“ â†“ â†“ â†“ â†“

import { BaseModel, __aggregationHeader } from "startupjs/orm";
export const access = undefined;
export const $$createdByUser = __aggregationHeader({
  collection: "games",
  name: "$$createdByUser",
});
export default class GamesModel extends BaseModel {
  async addNew() {
    await this.add({
      name: "New Game",
    });
  }
}


`;
